arrabiata/src/witness.rs

```diff
@@ -0,0 +1,1098 @@
+    // type-level
+    pub _marker: std::marker::PhantomData<(Fp, Fq, E1, E2)>,
+    // ---------------
+}
+
+// The condition on the parameters for E1 and E2 is to get the coefficients and
+// convert them into biguint.
+// The condition SWModelParameters is to get the parameters of the curve as
+// biguint to use them to compute the slope in the elliptic curve addition
+// algorithm.
+impl<
+        Fp: PrimeField,
+        Fq: PrimeField,
+        E1: CommitmentCurve<ScalarField = Fp, BaseField = Fq>,
+        E2: CommitmentCurve<ScalarField = Fq, BaseField = Fp>,
+    > InterpreterEnv for Env<Fp, Fq, E1, E2>
+where
+    <E1::Params as ark_ec::ModelParameters>::BaseField: PrimeField,
+    <E2::Params as ark_ec::ModelParameters>::BaseField: PrimeField,
+{
+    type Position = Column;
+
+    /// For efficiency, and for having a single interpreter, we do not use one
+    /// of the fields. We use a generic BigInt to represent the values.
+    /// When building the witness, we will reduce into the corresponding field.
+    // FIXME: it might not be efficient as I initially thought. We do need to
+    // make some transformations between biguint and bigint, with an extra cost
+    // for allocations.
+    type Variable = BigInt;
+
+    fn allocate(&mut self) -> Self::Position {
+        assert!(self.idx_var < NUMBER_OF_COLUMNS, "Maximum number of columns reached ({NUMBER_OF_COLUMNS}), increase the number of columns");
+        let pos = Column::X(self.idx_var);
+        self.idx_var += 1;
+        pos
+    }
+
+    fn allocate_public_input(&mut self) -> Self::Position {
+        assert!(self.idx_var_pi < NUMBER_OF_PUBLIC_INPUTS, "Maximum number of public inputs reached ({NUMBER_OF_PUBLIC_INPUTS}), increase the number of public inputs");
+        let pos = Column::PublicInput(self.idx_var_pi);
+        self.idx_var_pi += 1;
+        pos
+    }
+
+    fn write_column(&mut self, col: Self::Position, v: Self::Variable) -> Self::Variable {
+        let Column::X(idx) = col else {
+            unimplemented!("Only works for private inputs")
+        };
+        let modulus: BigInt = if self.current_iteration % 2 == 0 {
+            Fp::modulus_biguint().into()
+        } else {
+            Fq::modulus_biguint().into()
+        };
+        let v = v.mod_floor(&modulus);
+        self.state[idx] = v.clone();
+        v
+    }
+
+    fn write_public_input(&mut self, col: Self::Position, v: BigInt) -> Self::Variable {
+        let Column::PublicInput(idx) = col else {
+            unimplemented!("Only works for public input columns")
+        };
+        let modulus: BigInt = if self.current_iteration % 2 == 0 {
+            Fp::modulus_biguint().into()
+        } else {
+            Fq::modulus_biguint().into()
+        };
+        let v = v.mod_floor(&modulus);
+        self.public_state[idx] = v.clone();
+        v
+    }
+
+    /// Activate the gadget for the current row
+    fn activate_gadget(&mut self, gadget: Gadget) {
```

Comment:
this should not be in the witness computation right ?

Reply:
See my comment above about "setup environment".

Reply:
Agreed on the idea in a call.